### PR TITLE
Merge upstream v8.5.6

### PR DESCRIPTION
**Pull request by 2bndy5**
[patch_info.zip](https://github.com/jbms/sphinx-immaterial/files/9827216/patch_info.zip)

This wasn't as painful as I had expected (due to the long time since last update from upstream), or maybe I'm just getting better at it (wishful thinking 🤣 )

Aside from the usual JS tweaks after the merge script did it's thing, I also
- documented linked content tabs (addresses #61 )
- documented how to change the "Edit this page" icon

I had to tweak the merge script for files that are explicitly excluded (in `MKDOCS_EXCLUDE_PATTERNS`) and are no longer in the upstream base/target git indexes. There's still room for improvement because I had to manually add the new files from upstream since they weren't in this repo's index. The additions were mostly language translations and a couple features implemented on upstream only (like builtin site feedback and comments). Maybe there's a way to port the new functionality to sphinx builds, but I'm not overly enthusiastic about them (currently content with directing to other sites like github for a more dynamic user interactive experience).

I did discover 1 regression about mermaid graphs and filed an issue upstream (https://github.com/squidfunk/mkdocs-material/issues/4517). We can merge in the simple fix for that later though.

### CODE REVIEW

File: docs/content_tabs.rst
Comment (jbms) [inline docs/content_tabs.rst:108]: `navigation.instant` shouldn't be needed if this is working properly; if it doesn't, that suggests a problem with the use of localStorage --- also in my experience `navigation.instant` is kind of buggy and doesn't provide any significant benefit so I'm not sure we want to recommend it

File: src/assets/javascripts/_/index.ts
Comment (jbms) [inline src/assets/javascripts/_/index.ts:50]: Duplicate of below

Comment (jbms) [reply src/assets/javascripts/_/index.ts:50]: Probably should remove the one below instead to reduce merge conflicts in the future

File: src/assets/javascripts/_/index.ts
Comment (jbms) [inline src/assets/javascripts/_/index.ts:100]: we don't use these bug I guess they could be kept.  On the other hand it might be better to exclude them so that we catch at compile time any attempt to use them

File: src/assets/javascripts/_/index.ts
Comment (jbms) [inline src/assets/javascripts/_/index.ts:34]: are any of these other features working / should be documented?

File: src/assets/javascripts/components/toc/index.ts
Comment (jbms) [inline src/assets/javascripts/components/toc/index.ts:299]: Need to change `anchor.removeAttribute("data-md-state")` above to `anchor.classList.remove("md-nav__link--passed")`

File: src/assets/javascripts/templates/search/index.tsx
Comment (jbms) [inline src/assets/javascripts/templates/search/index.tsx:97]: I guess we could see about supporting tags?

File: src/assets/javascripts/_/index.ts
Comment (2bndy5) [reply src/assets/javascripts/_/index.ts:34]: I haven't tested them yet. I just figured it would hurt to have them while we figure it out.

File: src/assets/javascripts/components/toc/index.ts
Comment (2bndy5) [reply src/assets/javascripts/components/toc/index.ts:299]: good catch!

File: docs/content_tabs.rst
Comment (2bndy5) [reply docs/content_tabs.rst:108]: I quick copied and pasted the linked tabs example into a separate page and verified it worked with the feature specified in conf.py. I didn't think to try without it since the mkdocs-material doc mandated it be specified. I'll re-check without it explicitly enabled.

File: src/assets/stylesheets/main/layout/_nav.scss
Comment (jbms) [inline src/assets/stylesheets/main/layout/_nav.scss:117]: Not sure if this is correct --- will have to test after the other issue I mentioned related to `--passed` is fixed.

File: src/partials/feedback.html
Comment (jbms) [inline src/partials/feedback.html:30]: I guess we need to add a way to set this?

File: src/partials/footer.html
Comment (jbms) [inline src/partials/footer.html:30]: perhaps we need to add a way to set this?

File: src/partials/icons.html
Comment (jbms) [inline src/partials/icons.html:24]: kind of unfortunate implementation here since it is duplicated in every page rather than put in the css bundle, but I guess better than nothing...

File: src/partials/footer.html
Comment (2bndy5) [reply src/partials/footer.html:30]: Yeah, I think another meta data tag like our other `:hide-*:` mechanisms would suffice.

File: src/partials/feedback.html
Comment (2bndy5) [reply src/partials/feedback.html:30]: That would take me a while because I'm not well versed in site analytics. Here's hoping it is a simple patch to our current implementation of the analytics field in `html_theme_options`.

File: src/assets/stylesheets/main/layout/_nav.scss
Comment (2bndy5) [reply src/assets/stylesheets/main/layout/_nav.scss:117]: I allowed this because I noticed the change in [src/assets/javascripts/components/toc](https://github.com/jbms/sphinx-immaterial/pull/171/files#diff-ed6ae506600ba3c7de17fba6b64bc6a7f67a571894f3f73ce06ea04c63cb6b66). Are you specifically concerned about the `:not(&--in-viewport)`?

I didn't allow the change to the `toc.follow` feature (detailed in the patch_info.diff located in OP for this PR).

File: src/assets/javascripts/_/index.ts
Comment (jbms) [reply src/assets/javascripts/_/index.ts:34]: yeah sounds reasonable.

File: src/partials/feedback.html
Comment (jbms) [reply src/partials/feedback.html:30]: Sorry, I just meant a way to add it to page.meta.hide, e.g. via a `:hide-feedback:` meta option or something --- note the existing code in nav_adapt.py that modifies `page["meta"]["hide"]`

**Comment by jbms** [comment]
Thanks for doing this upgrade!

File: src/assets/javascripts/_/index.ts
Comment (2bndy5) [reply src/assets/javascripts/_/index.ts:100]: I'm ok with excluding them. If someone reports a problem, we can advise against using them or re-assess.

File: src/assets/stylesheets/main/layout/_nav.scss
Comment (jbms) [reply src/assets/stylesheets/main/layout/_nav.scss:117]: Yeah, I'm not sure if the other selectors can be removed or if they are still required --- will need to test once you have pushed out a fix for the other `--passed` change.

Comment (2bndy5) [reply src/assets/stylesheets/main/layout/_nav.scss:117]: pushed that `--passed` change.

**Review by jbms — APPROVED** [review]

**Comment by 2bndy5** [comment]
If you don't mind I'd like to keep this open for testing and further needed documentation about newer features. I'm about to push changes to docs about customizing the admonition icons. I also intend to patch the `hide-feedback` and `hide-footer` meta tags.

File: src/assets/stylesheets/main/layout/_nav.scss
Comment (2bndy5) [reply src/assets/stylesheets/main/layout/_nav.scss:117]: If you've approved, then I'm guessing this can be resolved?

**Comment by jbms** [comment]
> If you don't mind I'd like to keep this open for testing and further needed documentation about newer features. I'm about to push changes to docs about customizing the admonition icons. I also intend to patch the `hide-feedback` and `hide-footer` meta tags.

Sure, that's why I didn't merge it myself.

File: docs/content_tabs.rst
Comment (2bndy5) [reply docs/content_tabs.rst]: You're correct, `navigation.instant` isn't required for cross-page syncing of linked tabs. I have adjusted the doc here to express this difference from upstream.

Going forward, we should probably be careful about tab names. For example, a tab named "CSS hint" would be shown on page load if another tab of the same name was the last selected tab from a different section/page.

Comment (jbms) [reply docs/content_tabs.rst]: Agreed.  We also might wish to improve on the functionality provided by mkdocs-material to allow linking tabs more selectively, rather than an all or nothing option.  I also think in some cases it is used as an example for this theme in a way that might not be ideal for actual use.  My thinking is that tabs are best when there are true alternatives, where the user is unlikely to be interested in both options.  E.g. installation instructions for windows vs linux, or usage examples in Python vs C++.  If the user is likely to care about multiple tabs, then it may be better not to use tabs and instead just use sections.

File: src/assets/stylesheets/main/layout/_nav.scss
Comment (jbms) [reply src/assets/stylesheets/main/layout/_nav.scss:117]: yes, I verified that it works correctly

File: docs/content_tabs.rst
Comment (2bndy5) [reply docs/content_tabs.rst]: It is true that I tend to use tabs as a way of demonstrating steps in an example. My intention there was to keep the page length from growing to an overwhelming size -- I have a bad habit of using my mouse wheel more than the toc entries' links.

Maybe we should open an issue/discussion upstream about specifying certain tabs sets as non-sync-able?

File: src/assets/javascripts/_/index.ts
Comment (2bndy5) [reply src/assets/javascripts/_/index.ts:34]: `announce.dismiss` works well. I added a banner to our docs that declares the latest upstream version used (with a link to the upstream changelog). I also used the new HTML template as an example in the "New Blocks" section of the customization page. The `announce.dismiss` feature is enabled should we need to verify that it behaves consistently.

---
I'm not sure how to test the `content.lazy` feature since it is browser specific. So, I'm just leaving that flag in there in case we want to re-assess later.

File: src/assets/javascripts/templates/search/index.tsx
Comment (2bndy5) [reply src/assets/javascripts/templates/search/index.tsx:97]: I'm leaving this as is for now. If someone asks, then we can discuss it again (or just create a new issue as a reminder).

File: src/partials/feedback.html
Comment (2bndy5) [reply src/partials/feedback.html:30]: I think I found a bug (probably in Sphinx). The metadata keys don't get interpreted as metadata if the have a `-` in them. Rather, they just get interpreted as a field. I have tried all of our supported metadata tags, and they currently don't work. So, all our metadata keys should be changed to `hide_*` (thank goodnes we're still in beta).

I'm still digging, and I'll start rolling back the sphinx version I'm using (currently v5.3.0) to see where it was broken.

Comment (2bndy5) [reply src/partials/feedback.html:30]: Well the feedback can be displayed as a demo using a dummy analytics provider (like `G-XXXXXXXXXX`), but it would mean an additional warning in the browser's dev console. I've done this locally to verify that the new `:hide_feedback:` metadata tag works.

I started to look into doing this with my Google account, but apparently all traffic to analytics.google.com domain is blocked by my DNS sinkhole (PiHole) 😅 .

For now, I think I'll just put an example snippet in the docs under the analytics option in customization.rst.

**Comment by 2bndy5** [comment]
I'm now ready to merge this. I'll let it stew for a day so you have another chance to review the documentation changes (namely in customization.rst and admonitions.rst).

**Comment by 2bndy5** [comment]
The `consent` feature does work, but it's designed to not trigger on local files viewed in the browser. 
1. I'm not sure how to document the cookies portion of the feature.
2. I'm not sure the upstream default cookies are enacted (for google analytics and github) though.

Here's the config I used to test it:
```py
html_theme_options = {
    "consent": {
        "title": "Welcome!",
        "description": "We don't use cookies, so your privacy is safe here.",
        "actions": ["accept"],
    }
 }
```
and then manually remove the `hidden` attr in html because I was viewing it from a file.

I think I'll just add the necessary src code to use it and leave the documentation for another time.

File: src/partials/feedback.html
Comment (2bndy5) [reply src/partials/feedback.html:30]: > I'm still digging, and I'll start rolling back the sphinx version I'm using (currently v5.3.0) to see where it was broken.

Looks like the JSON domain relies on a module in sphinx that wasn't available in v4.3.0; we should probably update our requirements (or the optional deps for the json domain).

I couldn't find where it was broken. I tried rolling back sphinx (every version after v4.3.0) and docutils (as far back as v0.16), and the metadata tags weren't getting picked up with hyphens in the keys' name (using a fresh doc build each time).

Comment (jbms) [reply src/partials/feedback.html:30]: I'm seeing some confusing behavior with regard to the meta fields: I'm not able to reproduce the issue with hyphens in the meta fields, and indeed the Python and C++ apigen modules emit `:hide-edit-link:` which is correctly respected.  However, I am seeing that in some cases more than the first meta field is being ignored, but it will need more investigation to figure out what is going on.

For example, after adding `if docname == "index": print(meta)` to `_html_page_context` in `nav_adapt.py`, if I change the start of `docs/index.rst` to:

```
:foo-bar: abc def
:hero: A responsive Material Design theme for Sphinx sites.
```

Then both are present in meta:

```
{'foo-bar': 'abc def', 'hero': 'A responsive Material Design theme for Sphinx sites.'}
```

but if I reverse the order at the top of `docs/index.rst` to:

```
:hero: A responsive Material Design theme for Sphinx sites.
:foo-bar: abc def
```

The only `hero` is present:

```
{'hero': 'A responsive Material Design theme for Sphinx sites.'}
```

Comment (2bndy5) [reply src/partials/feedback.html:30]: Interesting, I didn't consider the order when diagnosing. I did try putting blank lines between the metadata in case the lack of a value associated with a metadata key may have been a problem.

Comment (2bndy5) [reply src/partials/feedback.html:30]: The only thing I can find in docutils changelogs is that they added a new directive for inserting [`<meta>`](https://docutils.sourceforge.io/docs/ref/doctree.html#meta) elements in v0.18. I also see they stopped stripping multiple `-` from CSS class names (which I remember having to contend with when porting one of upstream features to a theme ext).

Comment (jbms) [reply src/partials/feedback.html:30]: The problem seems to be related to having an `rst_prolog` specified in `conf.py`: the issue goes away when `rst_prolog` is removed.  It looks like the issue is in docutils but I'm still investigating.

Comment (jbms) [reply src/partials/feedback.html:30]: Actually the bug is in Sphinx.  When sphinx prepends the prolog, it attempts to insert it after the docinfo, which it detects using a regular expression:

https://github.com/sphinx-doc/sphinx/blob/93215f0a64a2e27656ff2bfe6958a20b1e6b6e1e/sphinx/util/rst.py#L76

https://github.com/sphinx-doc/sphinx/blob/93215f0a64a2e27656ff2bfe6958a20b1e6b6e1e/sphinx/util/rst.py#L26

This regular expression does not correctly match the full syntax allowed for fields for multiple reasons, but in particular it uses "\w" to match the field name, which excludes hyphens.

I think the intent here is to allow `rst_prolog` to insert content into the document (rather than just directives like `.. role: python(code)`), without preventing the document from specifying metadata fields.

For our `rst_prolog`, since it does not generate any docutils nodes, it makes no difference if it is inserted before or after the metadata fields.  The problem occurs if it is inserted in the middle of the metadata fields: since `:hero:` matches `docinfo_re` but `:foo-bar:` does not, the prolog gets inserted after `:hero:`, which results in the metadata fields being parsed as two separate field lists, and docutils only treats the first one as docinfo fields.

Comment (2bndy5) [reply src/partials/feedback.html:30]: Sounds like we should propose an amendment to regex in docutils.

Comment (jbms) [reply src/partials/feedback.html:30]: I filed this bug:
https://github.com/sphinx-doc/sphinx/issues/10932

As a workaround, leaving an initial blank line at the start of file before the docinfo fields mitigates the issue; since the regex doesn't match blank lines, the prolog will be inserted before it, and not split the docinfo fields.

Comment (2bndy5) [reply src/partials/feedback.html:30]: Ok I will revert the metadata key names back to using hyphens and test the workaround.

**Comment by jbms** [comment]
Thanks, looks good to me.

**Comment by 2bndy5** [comment]
Great, I'll merge this and start/continue squashing open issues. I suppose we could tag this merger, but I'm unreasonably excited about code annotations (despite the implementation's current limits), and want to solve #169 before the next PyPI release.

File: src/assets/stylesheets/main/layout/_nav.scss
Comment (2bndy5) [inline src/assets/stylesheets/main/layout/_nav.scss:681]: This `z-index` caused a regression as reported in #189

**Pull request merged**
